src/sentry/south_migrations/0456_backfill_groupenvironment_first_seen.py

```diff
@@ -0,0 +1,1278 @@
+# -*- coding: utf-8 -*-
+from south.utils import datetime_utils as datetime
+from south.db import db
+from south.v2 import DataMigration
+from django.db import models
+
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+
+
+class Migration(DataMigration):
+
+    # Flag to indicate if this migration is too risky
+    # to run online and needs to be coordinated for offline
+    is_dangerous = True
+
+    def forwards(self, orm):
+        db.commit_transaction()
+        try:
+            self._forwards(orm)
+        except Exception:
+            db.start_transaction()
+            raise
+        db.start_transaction()
+
+    def _forwards(self, orm):
+        "Write your forwards methods here."
+        GroupTagValue = orm.GroupTagValue
+        GroupEnvironment = orm.GroupEnvironment
+        Environment = orm.Environment
+
+        qs = GroupEnvironment.objects.filter(first_seen__isnull=True)
+        for ge in RangeQuerySetWrapperWithProgressBar(qs):
+            try:
+                env_name = Environment.objects.filter(
+                    id=ge.environment_id).values_list(
+                    'name', flat=True).get()
+            except Environment.DoesNotExist:
+                continue
+
+            try:
+                gtv = GroupTagValue.objects.get(
+                    group_id=ge.group_id,
+                    key='environment',
+                    value=env_name,
+                    first_seen__isnull=False,
+                )
+            except GroupTagValue.DoesNotExist:
+                continue
+
+            ge.first_seen = gtv.first_seen
+            ge.save()
```

Comment:
is this going to be fast enough? I was assuming we might have to create a table of the first_seen dates and do some kind of `UPDATE ... FROM` type trickery to batch it up a bit, but IDK, maybe this is fine as is.

Reply:
I imagine that'd be faster. Not sure long we're going to take to process 125MM rows serially, but it might be really slow.

How hard is it to get these sorts of tables in place? I assume we'd have to create them on default to be able to do the `UPDATE ... FROM`

Reply:
in the past, we've definitely done some that have taken hours to complete without issue, but i don't remember how many rows those were -- hoping @mattrobenolt can better estimate

Reply:
even if it takes 10ms to do the 2 roundtrips for the lookup and the save. it will take 14 days to do 125 million of them.

Reply:
lol fair

Reply:
Another option is to just partition the query on `GroupEnvironment` by id into X groups (maybe 10-20) then run them in parallel.